### PR TITLE
Test cases that fail

### PR DESCRIPTION
**Pull request by robstryker**
Hey, I was checking out your code. It looked great and well organized compared to the crap I was developing ;)  

But when I applied my test cases to yours, there are a ton of failures.  Almost all of my tests have been verified against the linux  'file' command, especially those dealing with simple numeric comparisons. It's always possible I have some gaps, but, it seems there's a lot of bad behavior here that would prevent your codebase from working with a magic file. 

One of the biggest problems is your treatment of signed comparisons. For example, a 4-byte comparison (ie, long) of  0xF0000000  compared to 0x00000000 should show 0x00000000 as being greater, because in a 4-byte signed comparison,  0xF0000000 is actually negative. 

But there's a lot of other stuff that fails, like pstring.  

This patch is just some suggested test cases, though.  I don't have any fixes yet, because I haven't figured out how to work within your architecture yet. I'll see if I can get around to patches, but I'm not sure if I will.

### CODE REVIEW

**Comment by j256** [comment]
Can you give me a specific example of a test that fails and show me the exception or output?  These tests work for me fine although I had a problem with the EXE discovery that I removed.

**Comment by j256** [comment]
Ok I get it.  Yeah it's screwed up.  I'm doing internal comparisons with a long but I guess an unsigned 32-bit value is expected.  I've got some test cases that fail and I'll incorporate yours.

**Comment by robstryker** [comment]
> I'm doing internal comparisons with a long but I guess an unsigned 32-bit value is expected. I've got some test cases that fail and I'll incorporate yours.

Actually you'll need to change it based on what's being compared.  For example, when comparing shorts in a signed fashion, you'll actually need to cast to short (as far as I can tell).  When doing shorts in an unsigned fashion, the comparison can be done inside the long and it won't matter.

**Comment by j256** [comment]
I'm not sure a cast is necessary but certainly the 2 complement stuff needs to be handled right.

**Comment by j256** [comment]
Ok.  I've copied and tweaked most of the tests into trunk by hand to better understand them and to do some optimization.  Hope that's okay @robstryker.  I wouldn't have taken the pull request anyway because the license at the top of the files was different from the library's anyway.

Thanks much for them though.  They helped me find problems with pstring and improper understanding of what the negation (~) operator does.  I'm still not 100% sure what to do with the 2's complement stuff.

**Comment by robstryker** [comment]
I ended up making my own solution (before finding yours, so don't take offense). Our code is similar but it wasn't so similar that I could simply make you a PR with the solution without significant chagnes to your files (which I know most people don't like). 

You can see one of my solutions here:   https://github.com/jbosstools/magicfile4j/blob/master/src/main/java/org/jboss/tools/magicfile4j/internal/model/matcher/ShortTest.java

Basically I keep track of whether it's signed or unsigned. I keep most data in a long structure (even when it's shorter than that).  For unsigned comparisons, I just use the long values to do the math. Otherwise, I down-cast them to their more appropriate type (short / int / etc)

**Comment by robstryker** [comment]
Oh, and no worries about taking the tests and modifying them. I'm all about open source. That's why i contributed them and offered them to you  ;)
